### PR TITLE
Added folding on header lines

### PR DESCRIPTION
**Pull request by WalrusGumboot**
When I find myself working on longer documents, I often enjoy maintaining a sense of structure by collapsing big blocks of the text at once. This PR adds some very basic functionality of doing so with folding ranges; it detects headings solely based on the '=' character. While this isn't ideal (something based on a syntactic analysis to fold all nodes that are a child of a certain heading would be more robust) it suits my personal needs well enough and might be worth considering.

### A quick note on the behaviour
In a document like
```typ
= this is a heading

some text

== subheading a

some more text

== subheading b

yet more text
```

folding ranges will be added as such:

```
┌──> │ = this is a heading
│    │ 
│    │ some text
└──< │ 
┌──> │ == subheading a
│    │ 
│    │ some more text
└──< │ 
┌──> │ == subheading b
│    │ 
└──< │ yet more text
```

instead of the hierarchical
```
┌──> │ = this is a heading
│    │ 
│    │ some text
│    │ 
│ ┌> │ == subheading a
│ │  │ 
│ │  │ some more text
│ └< │ 
│ ┌> │ == subheading b
│ │  │ 
│ └< │ yet more text
└──< │ 
```

### CODE REVIEW

File: src/server/folding_range.rs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+use super::TypstServer;
+
+impl TypstServer {
+    pub fn get_folding_ranges(&self, source: &Source) -> Option<Vec<FoldingRange>> {
+        let mut ranges = vec![];
+        let mut previous_idx: Option<usize> = None;
+
+        for (line_idx, line) in source.text().lines().enumerate() {
+            if line.starts_with('=') {
```
Comment (Enter-tainer) [inline src/server/folding_range.rs:12]: Can we get heading information using the AST? Detecting `==` is OK but fragile. The outline function also need to get all headings. This is how it is implemented: https://github.com/nvarner/typst-lsp/blob/master/src/server/symbols.rs#L98

Comment (astrale-sharp) [reply src/server/folding_range.rs:12]: We can and imo we should (i.e. raw text lines can start with `=`

Comment (WalrusGumboot) [reply src/server/folding_range.rs:12]: Ah, that's a good point, I hadn't considered it. I'll try my hand at using the AST's information later this evening ­— as much as I'd prefer not to, I've got exams to study for as well :sweat_smile:

Comment (WalrusGumboot) [reply src/server/folding_range.rs]: Okay, that was a lot easier than I thought it'd be. Should be all good now :)

**Comment by Myriad-Dreamin** [comment]
related #165

**Comment by Myriad-Dreamin** [comment]
@WalrusGumboot we are reaching a nice implementation, but unluckily I guess there is a bug when we simply use `document_symbol`. The result of `document_symbol` contains all symbols in a document. Therefore, in addition to headings, labels `<label>` and let binding `#let a = 1` and also separate a folding range here.

https://github.com/nvarner/typst-lsp/blob/4867da4b1e249182182549da6fb888fdf992ff48/src/server/symbols.rs#L34-L56

imo we should exclude them.

Could you exclude these unrelated symbols?

image

**Comment by WalrusGumboot** [comment]
Oops, didn't test extensively enough :sweat_smile: 

I guess filtering for `SymbolKind::NAMESPACE` works, and semantically it makes sense to fold on namespace-type symbols anyway.

**Comment by Myriad-Dreamin** [comment]
We have correct heading to folding now. But I find another bug, unfortunately. header foldings should be closed by their current parent, but now we don't and cannot do it simply. Accurately, a correct folding range identification should also regard brackets. Would you like to continue solving this inperfect? I personally think it doesn't bother much.

image

image

----

Btw, the default folding range on bracket are disabled since this PR.

image

**Comment by WalrusGumboot** [comment]
Hmm, yeah, that seems like more work than I've got time for at the moment. Maybe after the exam season.

**Comment by Myriad-Dreamin** [comment]
well, we may reserve this feature PR for you, @WalrusGumboot. You're free to reopen it later.

**Pull request closed without merging**